google/cloud/spanner/keys.h

```diff
@@ -0,0 +1,251 @@
+  explicit KeyRange(Bound<KeyType> start, Bound<KeyType> end)
+      : start_(start), end_(end) {}
+
+  /**
+   * Constructs a `KeyRange` closed on both `Bound`s.
+   * @param start
+   * @param end
+   */
+  explicit KeyRange(KeyType start, KeyType end)
+      : KeyRange(Bound<KeyType>(start), Bound<KeyType>(end)) {}
+
+  // Copy and move constructors and assignment operators.
+  KeyRange(KeyRange const& key_range) = default;
+  KeyRange& operator=(KeyRange const& rhs) = default;
+  KeyRange(KeyRange&& key_range) = default;
+  KeyRange& operator=(KeyRange&& rhs) = default;
+
+  bool IsStartClosed() const { return start_.IsClosed(); }
+  bool IsStartOpen() const { return start_.IsOpen(); }
+  bool IsEndClosed() const { return end_.IsClosed(); }
+  bool IsEndOpen() const { return end_.IsOpen(); }
+
+  KeyType const& StartKey() const { return start_.Key(); }
+  KeyType const& EndKey() const { return end_.Key(); }
+
+ private:
+  friend bool operator==(KeyRange const& lhs, KeyRange const& rhs) {
+    return lhs.start_ == rhs.start_ && lhs.end_ == rhs.end_;
+  }
+  friend bool operator!=(KeyRange const& lhs, KeyRange const& rhs) {
+    return !(lhs == rhs);
+  }
+
+  Bound<KeyType> start_;
+  Bound<KeyType> end_;
+};
+
+/**
+ * The `KeySet` class is a regular type that represents the collection of `Row`s
+ * necessary to uniquely identify a arbitrary group of rows in its index.
+ *
+ * A `KeySet` can consist of multiple `KeyRange` and/or `Row` instances.
+ */
+template <typename KeyType>
+class KeySet {
+ public:
+  /**
+   * Constructs an empty `KeySet`.
+   */
+  KeySet() = default;
+
+  /**
+   * Constructs a `KeySet` with a single key `Row`.
+   * @param key
+   */
+  explicit KeySet(KeyType key) : keys_(), key_ranges_() {
+    keys_.emplace_back(std::move(key));
+  }
+
+  /**
+   * Constructs a `KeySet` with a single `KeyRange`.
+   * @param key_range
+   */
+  explicit KeySet(KeyRange<KeyType> key_range) : keys_(), key_ranges_() {
+    key_ranges_.emplace_back(std::move(key_range));
+  }
+
+  /**
+   * Returns the key `Row`s in the `KeySet`.
+   * These keys are separate from the collection of `KeyRange`s.
+   */
+  std::vector<KeyType> const& keys() const { return keys_; }
+
+  /**
+   * Returns the `KeyRange`s in the `KeySet`.
+   * These are separate from the collection of individual key `Row`s.
+   */
+  std::vector<KeyRange<KeyType>> const& key_ranges() const {
+    return key_ranges_;
+  }
+
+  /**
+   * Adds a key `Row` to the `KeySet`.
+   *
+   * @param key
+   */
+  void Add(KeyType key) { keys_.emplace_back(std::move(key)); }
+
+  /**
+   * Adds a `KeyRange` to the `KeySet`.
+   *
+   * @param key_range
+   */
+  void Add(KeyRange<KeyType> key_range) {
+    key_ranges_.emplace_back(std::move(key_range));
+  }
+
+  /**
+   * Creates a key `Row` from the values provided and adds it to the `KeySet`.
+   *
+   * @tparam ValueTypes
+   * @param values
+   */
+  template <typename... ValueTypes>
+  void AddKey(ValueTypes... values) {
```

Comment:
I think we want to do one of the following:

a. Delete this method. Users can call `MakeRow` themselves
b. Name this `EmplaceKey` to indicate that it's *not* taking a key, it's taking the ingredients of a Key.

I prefer (a) because I don't think the slight savings of typing is worth the slight increase in API complexity.